### PR TITLE
OSCORE: Implements setting OSCORE information in endpoint contexts (remake)

### PR DESCRIPTION
**Pull request by rikard-sics**
This pull request creates functionality that adds strings with information about the OSCORE context used to the source or destination endpoint context for a CoAP message. These strings will be set in the endpoint context when communicating using OSCORE.

In certain situations an application may need to know OSCORE-related information about a CoAP message. For instance it may need to know which OSCORE Sender ID a message was transmitted from. The MapBasedEndpointContext contains a map of strings that can be used to add arbitrary information. This pull request makes it so that OSCORE sets some relevant strings in endpoint contexts when OSCORE is used. These strings can then be used by the application to retrieve information related to OSCORE from the endpoint context of a message.

See the new JUnit tests in this pull request for a demonstration of how this functionality can be used. Note that outgoing requests are handled in a special way with the new EndpointContextOperator functionality. For outgoing requests a new destination endpoint context will be created which is available via a callback using a MessageObserver. For outgoing responses the destination endpoint context used will be the source endpoint context created for the incoming request.

This pull request can help resolve the following issue:
https://github.com/eclipse/californium/issues/974
The previous approach of setting the PeerIdentity with a Principal for OSCORE (see https://github.com/eclipse/californium/pull/1002) was changed after discussion in that thread.

This is a new version of the previous PR on the same topic:
https://github.com/eclipse/californium/pull/1027

Other small changes in this pull request:
- Various formatting and comment fixes
- Moved flag indicating if responses will have a Partial IV to OSCoreCtx
- Set content format for error messages to text/plain
- Removed some code related to the old singleton from HashMapCtxDB

### CODE REVIEW

File: cf-oscore/src/main/java/org/eclipse/californium/oscore/HashMapCtxDB.java [hunk 1]
```diff
@@ -116,7 +109,9 @@ public synchronized void addContext(Token token, OSCoreCtx ctx) {
 	@Override
 	public synchronized void addContext(String uri, OSCoreCtx ctx) throws OSException {
 		if (uri != null) {
-			uriMap.put(normalizeServerUri(uri), ctx);
+		    String normalizedUri = normalizeServerUri(uri);
```
Comment (boaks) [inline cf-oscore/src/main/java/org/eclipse/californium/oscore/HashMapCtxDB.java]: My editor shows
\t\t\s\s\s\s
should be
\t\t\t

Do you like to reformat it? Or should we just merge it, "as it is"?

File: cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java [hunk 2]
```diff
@@ -79,8 +80,47 @@
...
 
-	//Include the context id in messages generated using this context
-	private boolean includeContextId;
+    /**
+     * Include the context id in messages generated using this context.
+     * This is generally optional and can be controlled by the application.
+     *
+     * Default value is false.
+     */
+    private boolean includeContextId;
```
Comment (boaks) [inline cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java]: My editor shows
\s\s\s\s
should be
\t

Do you like to reformat it? Or should we just merge it, "as it is"?

File: cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java [hunk 3]
```diff
@@ -79,8 +80,47 @@
...
+    /**
+     * String versions of the context ID, sender ID and recipient ID.
+     *
+     * These are set when the context is created (rather than for every message)
+     * to be used later when adding information about the messages to the
+     * EndpointContext on sending or receiving a message.
+     */
+    private final String contextIdString;
+    private final String senderIdString;
+    private final String recipientIdString;
```
Comment (boaks) [inline cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java]: My editor shows
\s\s\s\s
should be
\t

Do you like to reformat it? Or should we just merge it, "as it is"?

**Comment by boaks** [comment]
Beside the tabs/spaces, LGTM.
Do you want to go over it again and replace the spaces with tabs (apply the eclipse formatter)? I just marked the first one, the are more spaces.
Or should I just merge it?

**Comment by rikard-sics** [comment]
> Beside the tabs/spaces, LGTM.
> Do you want to go over it again and replace the spaces with tabs (apply the eclipse formatter)? I just marked the first one, the are more spaces.
> Or should I just merge it?

Sorry I only reformatted the EndpointContextInfoTest and OSCoreEndpointContextInfo, I missed to reformat some additions.

I will check the commit for spaces where tabs should be used and apply the formatter to the lines I am changing/adding.

**Comment by rikard-sics** [comment]
I have now applied the Californium formatter on added lines I could find with incorrect indentation. I could possibly do a future cleanup PR to make sure all code in the cf-oscore module is formatted with the Californium formatter.

Should I squash the commits in this PR?

File: cf-oscore/src/main/java/org/eclipse/californium/oscore/HashMapCtxDB.java (hunk 1 above)
Comment (rikard-sics) [reply cf-oscore/src/main/java/org/eclipse/californium/oscore/HashMapCtxDB.java]: Fixed in commit 6b98ac703fd398530455b8c5ea3f9e617268fed0

File: cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java (hunk 2 above)
Comment (rikard-sics) [reply cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java]: Fixed in commit 6b98ac703fd398530455b8c5ea3f9e617268fed0

File: cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java (hunk 3 above)
Comment (rikard-sics) [reply cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java]: Fixed in commit 6b98ac703fd398530455b8c5ea3f9e617268fed0

**Comment by boaks** [comment]
Yes, please squash your commit and I will merge it afterwards!
Thanks!

**Comment by rikard-sics** [comment]
> Yes, please squash your commit and I will merge it afterwards!
> Thanks!

Sounds good, the commits have now been squashed. Thanks for all the assistance.

**Pull request merged**